Modules/src/UserInputValidator.cc

```diff
@@ -44,6 +61,86 @@ namespace ChimeraTK {
     return rejected;
   }
 
+  void UserInputValidator::enableDeepValidation() {
+    // Find out accessors with return
+    for(auto& accessor : _module->getAccessorListRecursive()) {
+      if(accessor.getDirection() == VariableDirection{VariableDirection::feeding, true} &&
+          accessor.getModel().getTags().count(std::string(UserInputValidator::tagValidatedVariable)) > 0) {
+        _downstreamInvalidatingReturnChannels.emplace(accessor.getAppAccessorNoType().getId());
+      }
+    }
+
+    auto keepPvAccesWithReturnChannel = Model::EdgeFilter([](const Model::EdgeProperties& edge) -> bool {
+      return edge.type == Model::EdgeProperties::Type::pvAccess && edge.pvAccessWithReturnChannel;
+    });
+
+    // Filter that keeps our tagged variable nodes and application modules
+    auto validatedVariablesAndApplicationModulesFilter =
+        Model::VertexFilter([](const Model::VertexProperties& vertex) -> bool {
+          return vertex.visit([](auto props) -> bool {
+            if constexpr(Model::isVariable(props)) {
+              return props.tags.count(std::string(UserInputValidator::tagValidatedVariable)) > 0;
+            }
+
+            if constexpr(Model::isApplicationModule(props)) {
+              return true;
+            }
+
+            return false;
+          });
+        });
+
+    // Find longest path that is validated in our model, starting at this module
+    std::deque<Model::ApplicationModuleProxy> stack;
+    std::map<ApplicationModule*, int> distances;
+    distances[_module] = 0;
+
+    auto orderVisitor = [&](auto proxy) {
+      if constexpr(Model::isApplicationModule(proxy)) {
+        stack.push_front(proxy);
+        distances.try_emplace(&proxy.getApplicationModule(), std::numeric_limits<int>::min());
+      }
+    };
+
+    _module->getModel().visit(orderVisitor, Model::visitOrderPost, Model::depthFirstSearch,
+        keepPvAccesWithReturnChannel, validatedVariablesAndApplicationModulesFilter);
```

Comment:
I fear this does not do what you intend. `visit()` will always iterate all vertices in the model that are reachable with the given edge filter, independent of the vertex filter. The vertex filter will only control for which found vertices the visitor functor is called. The vertex filter does not prevent the DFS to continue a certain path. Only the edge filter does that.

Since you have `if constexpr(Model::isApplicationModule(proxy))` in your visitor functor, the vertex filter is completely redundant. You can leave out and you will get the same result here.

As a consequence, this DFS will not stop in the hierarchy at the lowest level of validators. As long as there are return channels, it will continue its search.

Maybe this is actually ok, since we only want to estimate the maximum possible validation depth here. Return channels are not so common, and it might actually be that some other validation mechanism is behind the return channel. This code can be quite simplified then though, and we should mention in a comment that this is an approximation.